### PR TITLE
Input Interaction: better horizontal edge detection

### LINKED ISSUE #12114: Tables block - BUG: Arrow keys not working as expected...
### Describe the bug
**In the Tables block:**
- DOWN arrow doesn't do anything (should move down a block)
- UP arrow doesn't move up (instead, it moves left, only every-other-time)
- LEFT arrow button only works in empty cells

Gutenberg 4.4

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #12114.

Fixes horizontal arrow navigation in empty containers. To reproduce, create an empty paragraph block and try to use the left/right arrow keys.

This completely rewrites `isHorizontalEdge` to check visual position instead of DOM position, just like `isVerticalEdge`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by ellatrix** [comment]
All current test in `master` pass. I'm trying to make a test specifically for the empty paragraph node, and while it works consistently testing manually, for some reason it doesn't work in Puppeteer. I'll need to investigate at another time.

**Comment by nosolosw** [comment]
Tried this patch on an empty paragraph and a table (as in the original issue):

- on empty paragraph:
  - Chrome: works as expected.
  - Firefox: arrow up doesn't work after the first time.

- on the table block:
  - Chrome: left/right work as expected; down works as right, up works as left.
  - Firefox: left/right work as expected; down works as right, up doesn't work on empty cells (in cell with content works as left).

As for the down/up keys: I've tested tables in office applications (writer, spreadsheets). There, the arrow down moves to the cell below instead of to the right cell (and the arrow up to the cell above instead of the left cell). Although I think that would be the desirable behavior, this is great progress and should be fine to track that in a different issue if you think it's better.

Tested on Firefox 66 and Chrome 71, Linux OS.

**Comment by ellatrix** [comment]
@nosolosw Thanks for testing. It's definitely not easy to fix it for Firefox, as it totally mispositions the caret rectangle sometimes. I'll see if I can make a more reliable calculation.

Regarding the up and down arrows working differently in a table, I think this is a separate issue, but worth fixing in a separate PR. The problem is that we first look in the DOM for a tabbable element, then calculate the new caret position.

**Comment by ellatrix** [comment]
But that confirms again that we shouldn't be looking in the DOM when repositioning the caret with arrow up/down. ;)

**Comment by ellatrix** [comment]
Maybe these functions are also in the wrong package... :/

**Comment by ellatrix** [comment]
@nosolosw I fixed vertical edge as well, and also merged the logic.

File: packages/dom/src/dom.js [hunk 1]
```diff
@@ -60,14 +60,17 @@ function isSelectionForward( selection ) {
...
 /**
- * Check whether the selection is horizontally at the edge of the container.
+ * Check whether the selection is at the edge of the container. Checks for
+ * horizontal position by default. Set `isVertical` to true to check only
+ * vertically.
  *
- * @param {Element} container Focusable element.
- * @param {boolean} isReverse Set to true to check left, false for right.
+ * @param {Element} container  Focusable element.
+ * @param {boolean} isReverse  Set to true to check left, false to check right.
```
Comment (nosolosw) [inline packages/dom/src/dom.js]: This should [be](http://usejsdoc.org/tags-param.html#parameters-with-properties):

```
@param {Element} container ...
@param {Object} opts ...
@param {Object} opts.isReverse ...
@param {Object} opts.isVertical ...
```

File: packages/dom/src/dom.js [hunk 2]
```diff
@@ -60,14 +60,17 @@ function isSelectionForward( selection ) {
...
+ * @param {boolean} isVertical Set to true to check only vertical position.
  *
  * @return {boolean} True if at the horizontal edge, false if not.
```
Comment (nosolosw) [inline packages/dom/src/dom.js]:
```suggestion
 * @return {boolean} True if at the edge, false if not.
```

File: packages/dom/src/dom.js [hunk 3]
```diff
@@ -198,20 +112,60 @@ export function isVerticalEdge( container, isReverse ) {
...
+	return isEdge( container, { isReverse } );
+}
+
+/**
+ * Check whether the selection is vertically at the edge of the container.
+ *
+ * @param {Element} container Focusable element.
+ * @param {boolean} isReverse Set to true to check top, false for bottom.
+ *
+ * @return {boolean} True if at the edge, false if not.
```
Comment (nosolosw) [inline packages/dom/src/dom.js]:
```suggestion
 * @return {boolean} True if at the vertical edge, false if not.
```

File: packages/dom/src/dom.js [hunk 4]
```diff
@@ -198,20 +112,60 @@ export function isVerticalEdge( container, isReverse ) {
...
 
-	const editableRect = container.getBoundingClientRect();
-	const buffer = lineHeight / 2;
+	// Calculate a buffer that is half the line height. In some browsers, the
+	// selection rectangle may not fill the entire height of the line, so we add
+	// 3/4 the line height to the selection rectangle to ensure that it is well
+	// over its line boundary.
+	const buffer = 3 * parseInt( lineHeight, 10 ) / 4;
+	const containerRect = container.getBoundingClientRect();
+	const verticalEdge = isReverse ?
```
Comment (nosolosw) [inline packages/dom/src/dom.js]: How about renaming this variable as `isAtEdge`?

File: packages/dom/src/dom.js [hunk 5]
```diff
@@ -60,14 +60,17 @@ function isSelectionForward( selection ) {
...
- * @param {Element} container Focusable element.
- * @param {boolean} isReverse Set to true to check left, false for right.
+ * @param {Element} container  Focusable element.
+ * @param {boolean} isReverse  Set to true to check left, false to check right.
+ * @param {boolean} isVertical Set to true to check only vertical position.
  *
  * @return {boolean} True if at the horizontal edge, false if not.
  */
-export function isHorizontalEdge( container, isReverse ) {
+function isEdge( container, { isReverse, isVertical } ) {
```
Comment (nosolosw) [inline packages/dom/src/dom.js]: How about renaming `isVertical` to `onlyVertical`? My brain was confused by this name and somehow assumed it worked the same as `isReverse` (check vertical or horizontal).

File: packages/dom/src/dom.js [hunk 6]
```diff
@@ -198,20 +112,60 @@ export function isVerticalEdge( container, isReverse ) {
...
+	if ( ! verticalEdge ) {
 		return false;
 	}
 
-	// Too high.
-	if ( ! isReverse && rangeRect.bottom + buffer < editableRect.bottom ) {
+	if ( isVertical ) {
+		return true;
+	}
+
```
Comment (nosolosw) [inline packages/dom/src/dom.js]: If I understood this trick correctly, we create a fake range positioned to the edges of the container as to create a rectangle that bounds the text: thanks to this we can detect void elements or text not stretched to the edge of the container. Is that right? Are there any more use cases? I guess this merits a comment :)

Comment (ellatrix) [reply packages/dom/src/dom.js]: Fixed in ccce7e0.

File: packages/dom/src/dom.js (hunk 5 above)
Comment (ellatrix) [reply packages/dom/src/dom.js]: Fixed in ccce7e0.

File: packages/dom/src/dom.js (hunk 3 above)
Comment (ellatrix) [reply packages/dom/src/dom.js]: Fixed in ccce7e0.

File: packages/dom/src/dom.js (hunk 2 above)
Comment (ellatrix) [reply packages/dom/src/dom.js]: Fixed in ccce7e0.

File: packages/dom/src/dom.js (hunk 1 above)
Comment (ellatrix) [reply packages/dom/src/dom.js]: Fixed in ccce7e0.

**Comment by ellatrix** [comment]
Thanks for reviewing @nosolosw! I addressed your feedback.

File: packages/dom/README.md [hunk 7]
```diff
@@ -148,7 +148,7 @@ _Parameters_
 
 _Returns_
 
--   `boolean`: True if at the edge, false if not.
+-   `boolean`: True if at the vertical edge, false if not.
```
Comment (nosolosw) [inline packages/dom/README.md]: isn't it lovely that we don't have to remember to update this? :)

**Review by oandregal — APPROVED** [review]
This makes keyboard navigation a pleasure! :star2:

**Comment by nosolosw** [comment]
Created https://github.com/WordPress/gutenberg/issues/14675 for the UP/DOWN navigation within a table block.

**Comment by ellatrix** [comment]
Thanks for your review! ✨

**Pull request merged**